### PR TITLE
Elasticsearch Retry Config

### PR DESCRIPTION
**Pull request by KatrinIhler**
This adds configuration options to Elasticsearch for retrying failed requests. This is helpful if the index is temporarily unavailable, e.g. if Elasticsearch doesn't have enough heap to fulfill the request. You can set how many times a retry will be attempted, as well as how long to wait between attempts. These can be configured separately for simple get requests and for requests that also update or delete, since the latter are usually more important. Note that a get request that is necessary for an update will be considered part of the update, so those config settings apply.

Additionally I cleaned up some code (mostly recommendations from Intellij), and also made the index name (not identifier) changeable during runtime. Updates to the documentation and text for the release notes are included.

@wsmirnow asked me to aim this at OC 10, but as this depends on #2857, that's unfortunately not possible. Not sure if OC 11 is the right branch, since it isn't out yet, or if this would be better aimed at develop because this adds functionality to existing code. Any opinions?

### CODE REVIEW

**Comment by mliradelc** [comment]
Hi Kathrin, @joneug and me as release Managers of version 11 we decided to include your PR into opencast 11, can you target to `r/11.x`?

**Comment by KatrinIhler** [comment]
> Hi Kathrin, @joneug and me as release Managers of version 11 we decided to include your PR into opencast 11, can you target to `r/11.x`?

Already am. :)

File: etc/org.opencastproject.elasticsearch.index.ElasticsearchIndex.cfg
Comment (gregorydlogan) [inline etc/org.opencastproject.elasticsearch.index.ElasticsearchIndex.cfg]: I'll fix these after merge, but IMO defaults should match verbatim with the setting below it.  You and I can parse None == 0, but someone who *isn't* as fluent might not be able to.

File: modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java
Comment (gregorydlogan) [inline modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java:160]: Out of curiosity, why are these here rather than in the superclass?

File: modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java
Comment (gregorydlogan) [inline modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java:162]: I have to be that guy, but `max.waiting.*=-1` does bad things since Thread.sleep throws an `IllegalArgumentException` for a negative waiting time.

Further objections: While *technically* correct in that a maximum retry count of -1 never retries, I'm pretty sure we should throw an exception, or at least warn that this is a dumb idea.  Also, setting non-integer values yeilds 0s, but I'm not sure there's much you can do with that since you're using NumberUtils.

**Review by gregorydlogan — CHANGES_REQUESTED** [review]

**Comment by mliradelc** [comment]
Last or two weeks ago, we talked in the dev meeting about this PR to be delayed for the next minor release. I will leave as a draft

File: modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java
Comment (KatrinIhler) [reply modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java:160]: The reasoning was that properties that are used by the logic of the super class will get read in by that one, while the others will be left to the sub class. The sub class decides which configured retry attempts/waiting period are passed over to the super class as method parameters depending on if it's a read or update, which the super class doesn't know about, so I figured these would be mainly the sub class's responsibility. 

But let's be honest, the whole separation between those two classes is a bit messy anyway. I tried to straighten it out a bit in the last few PRs, but... yeah.

File: modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java
Comment (KatrinIhler) [reply modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java:162]: The code actually checks for negative waiting periods:
```
          if (retryWaitingPeriod > 0) {
            Thread.sleep(retryWaitingPeriod);
          }
```

But I'm gonna add a warning for negative configured values.

The waiting period is in milliseconds, that should be enough granularity for people... :laughing:

**Comment by KatrinIhler** [comment]
@gregorydlogan Finally addressed all your findings, ready for review again :)

**Pull request merged**
